File: src/lib/tiposkripto/BaseGiven.ts
import type { Ibdd_in_any } from "../CoreTypes";
import { ITestArtifactory, ITestResourceConfiguration } from ".";

/**
 * Represents a collection of Given conditions keyed by their names.
 * Givens are typically organized as named collections because:
 * - They set up different initial states for tests
 * - Tests often need to reference specific Given conditions by name
 * - This allows for better organization and reuse of setup logic
 * - The BDD pattern often involves multiple named Given scenarios
 */
export type IGivens<I extends Ibdd_in_any> = Record<string, BaseGiven<I>>;

export abstract class BaseGiven<I extends Ibdd_in_any> {
  features: string[];
  whens: any[];
  thens: any[];
  error: Error;
  fail: any;
  store: I["istore"];
  recommendedFsPath: string;
  givenCB: I["given"];
  initialValues: any;
  key: string;
  failed: boolean;
  artifacts: string[] = [];

  status: boolean | undefined;

  addArtifact(path: string) {
    if (typeof path !== "string") {
      throw new Error(
        `[ARTIFACT ERROR] Expected string, got ${typeof path}: ${JSON.stringify(
          path
        )}`
      );
    }
    const normalizedPath = path.replace(/\\/g, "/"); // Normalize path separators
    this.artifacts.push(normalizedPath);
  }

  constructor(
    features: string[],
    whens: any[],
    thens: any[],
    givenCB: I["given"],
    initialValues: any
  ) {
    this.features = features;
    this.whens = whens;
    this.thens = thens;
    this.givenCB = givenCB;
    this.initialValues = initialValues;
    this.fails = 0; // Initialize fail count
  }

  beforeAll(store: I["istore"]) {
    return store;
  }

  toObj() {
    return {
      key: this.key,
      whens: (this.whens || []).map((w) => {
        if (w && w.toObj) return w.toObj();
        console.error("When step is not as expected!", JSON.stringify(w));
        return {};
      }),
      thens: (this.thens || []).map((t) => (t && t.toObj ? t.toObj() : {})),
      error: this.error ? [this.error, this.error.stack] : null,
      failed: this.failed,
      features: this.features || [],
      artifacts: this.artifacts,
      status: this.status,
    };
  }

  abstract givenThat(
    subject: I["isubject"],
    testResourceConfiguration,
    artifactory: ITestArtifactory,
    givenCB: I["given"],
    initialValues: any
  ): Promise<I["istore"]>;

  async afterEach(
    store: I["istore"],
    key: string,
    artifactory: ITestArtifactory
  ): Promise<I["istore"]> {
    return store;
  }

  async give(
    subject: I["isubject"],
    key: string,
    testResourceConfiguration: ITestResourceConfiguration,
    tester: (t: Awaited<I["then"]> | undefined) => boolean,
    artifactory: ITestArtifactory,
    suiteNdx: number
  ) {
    this.key = key;
    this.fails = 0; // Initialize fail count for this given

    const givenArtifactory = (fPath: string, value: unknown) =>
      artifactory(`given-${key}/${fPath}`, value);

    try {
      // Ensure addArtifact is properly bound to 'this'
      const addArtifact = this.addArtifact.bind(this);

      this.store = await this.givenThat(
        subject,
        testResourceConfiguration,
        givenArtifactory,
        this.givenCB,
        this.initialValues
      );
      this.status = true;
    } catch (e) {
      this.status = false;

      this.failed = true;
      this.fails++; // Increment fail count
      this.error = e.stack;
    }

    try {
      const whens = this.whens || [];
      for (const [thenNdx, thenStep] of this.thens.entries()) {
        try {
          const t = await thenStep.test(
            this.store,
            testResourceConfiguration,
            `suite-${suiteNdx}/given-${key}/then-${thenNdx}`
          );
          // If the test doesn't throw, it passed
          tester(t);
        } catch (e) {
          // Mark the given as failed if any then step fails
          this.failed = true;
          this.fails++; // Increment fail count
          // Re-throw to propagate the error
          throw e;
        }
      }
    } catch (e) {
      this.error = e.stack;
      this.failed = true;
    } finally {
      try {
        const addArtifact = this.addArtifact.bind(this);
        await this.afterEach(this.store, this.key);
      } catch (e) {
        this.failed = true;
        this.fails++; // Increment fail count
        throw e;
        // this.error = e.message;
      }
    }

    return this.store;
  }
}

File: src/lib/rubeno/base_given.rb
module Rubeno
  class BaseGiven
    attr_accessor :features, :whens, :thens, :error, :fail, :store, :recommended_fs_path,
                  :given_cb, :initial_values, :key, :failed, :artifacts, :status, :fails
    
    def initialize(features, whens, thens, given_cb, initial_values)
      @features = features
      @whens = whens
      @thens = thens
      @given_cb = given_cb
      @initial_values = initial_values
      @artifacts = []
      @fails = 0
    end
    
    def add_artifact(path)
      normalized_path = path.gsub('\\', '/')
      @artifacts << normalized_path
    end
    
    def before_all(store)
      store
    end
    
    def to_obj
      {
        key: @key,
        whens: (@whens || []).map { |w| w.respond_to?(:to_obj) ? w.to_obj : {} },
        thens: (@thens || []).map { |t| t.respond_to?(:to_obj) ? t.to_obj : {} },
        error: @error ? [@error, @error.backtrace] : nil,
        failed: @failed,
        features: @features || [],
        artifacts: @artifacts,
        status: @status
      }
    end
    
    def given_that(subject, test_resource_configuration, artifactory, given_cb, initial_values, pm)
      raise NotImplementedError, "given_that must be implemented by subclasses"
    end
    
    def after_each(store, key, artifactory, pm)
      store
    end
    
    def give(subject, key, test_resource_configuration, tester, artifactory, t_log, pm, suite_ndx)
      @key = key
      @fails = 0
      @failed = false
      @error = nil
      
      given_artifactory = ->(f_path, value) do
        artifactory.call("given-#{key}/#{f_path}", value)
      end
      
      begin
        # Call the given callback to get initial subject
        initial_subject = @given_cb.call(@initial_values)
        @store = given_that(
          initial_subject,
          test_resource_configuration,
          given_artifactory,
          @given_cb,
          @initial_values,
          pm
        )
        @status = true
      rescue => e
        @status = false
        @failed = true
        @fails += 1
        @error = e
        return @store
      end
      
      begin
        # Process whens
        @whens.each_with_index do |when_step, when_ndx|
          begin
            @store = when_step.test(
              @store,
              test_resource_configuration,
              t_log,
              pm,
              "suite-#{suite_ndx}/given-#{key}/when/#{when_ndx}"
            )
          rescue => e
            @failed = true
            @fails += 1
            @error = e
          end
        end
        
        # Process thens
        @thens.each_with_index do |then_step, then_ndx|
          begin
            result = then_step.test(
              @store,
              test_resource_configuration,
              t_log,
              pm,
              "suite-#{suite_ndx}/given-#{key}/then-#{then_ndx}"
            )
            unless tester.call(result)
              @failed = true
              @fails += 1
            end
          rescue => e
            @failed = true
            @fails += 1
            @error = e
          end
        end
      rescue => e
        @error = e
        @failed = true
        @fails += 1
      ensure
        begin
          after_each(@store, @key, given_artifactory, pm)
        rescue => e
          @failed = true
          @fails += 1
        end
      end
      
      @store
    end
  end
end
